module/zfs/dmu.c

```diff
@@ -851,7 +893,7 @@ dmu_free_long_range_impl(objset_t *os, dnode_t *dn, uint64_t offset,
 
 		mutex_enter(&dp->dp_lock);
 		dp->dp_long_free_dirty_pertxg[dmu_tx_get_txg(tx) & TXG_MASK] +=
-		    chunk_len;
+		    chunk_data_count;
```

Comment:
Let's add a comment here to explain what's going on, and why we're not using `chunk_len`.
I think we can also drop the local variable to make it even clearer we only need this value in
one place.  How about something like this:

```c
                /*
                 * In order to prevent unnecessary TXG throttling only the
                 * segments of the chunk being freed which contain data
                 * should be counted toward the per-TXG dirty limit.
                 */
                mutex_enter(&dp->dp_lock);
                dp->dp_long_free_dirty_pertxg[dmu_tx_get_txg(tx) & TXG_MASK] +=
                   dmu_dnode_count_data_in_chunk(dn, chunk_begin, chunk_len);
                mutex_exit(&dp->dp_lock);
```